File: mockCompiler/src/main/java/com/azazellj/test/mockcompiler/MockCompilerProcessor.java
package com.azazellj.test.mockcompiler;

import com.azazellj.mock.annotations.Mock;
import com.azazellj.mock.annotations.MockTypeHelper;
import com.azazellj.mock.annotations.internal.MockUtils;
import com.azazellj.test.mockcompiler.generator.Helper;
import com.azazellj.test.mockcompiler.generator.MockGenerator;
import com.google.auto.service.AutoService;

import javax.annotation.processing.*;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.MirroredTypesException;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import javax.tools.Diagnostic;
import java.io.IOException;
import java.util.*;

import static com.google.common.base.Preconditions.checkNotNull;

@AutoService(Processor.class)
public class MockCompilerProcessor extends AbstractProcessor {

    Filer mFiler;
    Messager mMessager;
    Elements mElementUtils;
    Types mTypes;

    private Map<String, MockAnnotatedClass> mockClasses = new LinkedHashMap<>();
    public static Map<String, String> registeredFunctions = new HashMap<>();

    @Override
    public synchronized void init(ProcessingEnvironment env) {
        super.init(env);

        mFiler = env.getFiler();
        mMessager = env.getMessager();
        mElementUtils = env.getElementUtils();
        mTypes = env.getTypeUtils();
    }


    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment env) {
        registerDefaultHelpers();
        registerCustomHelpers(env);

        findAnnotatedClasses(env);
        registerHelpersFromMockAnnotatedClasses();

        generateMockClasses();

        return false;
    }

    private void generateMockClasses() {
        try {
            for (MockAnnotatedClass mockClass : mockClasses.values()) {
                MockGenerator.generateClass(mElementUtils,mTypes,mFiler, mMessager, mockClass);
            }
        } catch (IOException ioEx) {
            error(null, ioEx.toString());
        }

        mockClasses.clear();
    }

    private void findAnnotatedClasses(RoundEnvironment env) {
        Set<? extends Element> elements = env.getElementsAnnotatedWith(Mock.class);

        for (Element annotatedElement : elements) {
            if (!isValidAnnotatedElement(annotatedElement)) continue;

            registerMockAnnotatedClass(annotatedElement);
        }
    }

    private void registerHelpersFromMockAnnotatedClasses() {
        for (String preGeneratedClassName : mockClasses.keySet()) {
            String fullClassName = mockClasses.get(preGeneratedClassName).getAnnotatedClassElement().toString();
            String postGeneratedFunction = Helper.fullGeneratedHelperPath(preGeneratedClassName);

            registeredFunctions.put(fullClassName, postGeneratedFunction);
        }
    }

    private void registerMockAnnotatedClass(Element element) {
        MockAnnotatedClass annotatedClass = new MockAnnotatedClass(element);
        String elementName = elementName(element);

        if (mockClasses.containsKey(elementName)) {
            // TODO: 1/7/18 write overriding and prioritizing
        }

        mockClasses.put(elementName, annotatedClass);
    }

    private boolean isValidAnnotatedElement(Element element) {
        boolean isValid = true;
        if (element.getKind() != ElementKind.CLASS) {
            error(element, "Only classes can be annotated with @%s", Mock.class.getSimpleName());
            isValid = false;
        }
        if (element.getModifiers().contains(Modifier.PRIVATE)) {
            error(element, "Abstract class can`t be annotated with @%s", Mock.class.getSimpleName());
            isValid = false;
        }

        return isValid;
    }

    private void registerDefaultHelpers() {
        TypeElement typeElement = mElementUtils.getTypeElement(MockUtils.class.getName());
        registerFunctions(typeElement.getEnclosedElements());
    }

    private void registerCustomHelpers(RoundEnvironment env) {
        registerFunctions(env.getElementsAnnotatedWith(MockTypeHelper.class));
    }

    void registerFunctions(Collection<? extends Element> elements) {
        for (Element element : elements) {
            if (!isMockHelperMethodValid(element)) continue;

            String helperFunctionName = helperFunctionName(element);
            registerFunction(element, helperFunctionName);
        }
    }

    void registerFunction(Element element, String helperFunctionName) {
        try {
            MockTypeHelper helper = element.getAnnotation(MockTypeHelper.class);
            for (Class supportedClass : helper.types()) {
                // TODO: 1/6/18 write overriding and prioritizing
                registeredFunctions.put(supportedClass.toString(), helperFunctionName);
            }
        } catch (MirroredTypesException typesEx) {
            for (TypeMirror typeMirror : typesEx.getTypeMirrors()) {
                // TODO: 1/6/18 write overriding and prioritizing
                registeredFunctions.put(typeMirror.toString(), helperFunctionName);
            }
        }

//        mMessager.printMessage(Diagnostic.Kind.ERROR, helperFunctionName);


    }


    private void error(Element e, String msg, Object... args) {
        mMessager.printMessage(
                Diagnostic.Kind.ERROR,
                String.format(msg, args),
                e);
    }

    @Override
    public Set<String> getSupportedAnnotationTypes() {
        Set<String> annotations = new HashSet<>();
        annotations.add(Mock.class.getCanonicalName());
        annotations.add(MockTypeHelper.class.getCanonicalName());
        return annotations;
    }

    @Override
    public SourceVersion getSupportedSourceVersion() { return SourceVersion.latestSupported();}

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // HELP METHODS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    String elementName(Element element) {
        return checkNotNull(element).getSimpleName().toString();
    }

    boolean isMockHelperMethodValid(Element element) {
        checkNotNull(element);

        boolean isMethod = element.getKind() == ElementKind.METHOD;
        boolean isStatic = element.getModifiers().contains(Modifier.STATIC);
        boolean isPublic = element.getModifiers().contains(Modifier.PUBLIC);
        return isMethod && isPublic && isStatic;
    }

    String helperFunctionName(Element element) {
        checkNotNull(element);
        checkNotNull(element.getEnclosingElement());

        String className = element.getEnclosingElement().asType().toString();
        String functionName = element.toString();
        return String.format("%s.%s", className, functionName);
    }
}

File: mockCompiler-kt/src/main/kotlin/com/azazellj/test/mockcompiler/MockCompilerProcessor.kt
package com.azazellj.test.mockcompiler

import com.azazellj.mock.annotations.Mock
import com.azazellj.mock.annotations.MockTypeHelper
import com.azazellj.mock.annotations.internal.MockUtils
import com.azazellj.test.mockcompiler.generator.Helper
import com.azazellj.test.mockcompiler.generator.MockGenerator
import javax.annotation.processing.*
import javax.lang.model.SourceVersion
import javax.lang.model.element.Element
import javax.lang.model.element.ElementKind
import javax.lang.model.element.Modifier
import javax.lang.model.element.TypeElement
import javax.lang.model.type.MirroredTypesException
import javax.lang.model.util.Elements
import javax.lang.model.util.Types
import javax.tools.Diagnostic

@AutoService(Processor::class)
class MockCompilerProcessor : AbstractProcessor() {

    companion object {
        val registeredFunctions: MutableMap<String, String> = mutableMapOf()
    }

    lateinit var mFiler: Filer
    lateinit var mMessager: Messager
    lateinit var mElementUtils: Elements
    lateinit var mTypes: Types

    private val mockClasses: MutableMap<String, MockAnnotatedClass> = mutableMapOf()


    override fun init(env: ProcessingEnvironment) {
        super.init(env)

        mFiler = env.filer
        mMessager = env.messager
        mElementUtils = env.elementUtils
        mTypes = env.typeUtils
    }


    override fun process(annotations: MutableSet<out TypeElement>, env: RoundEnvironment): Boolean {
        registerDefaultHelpers()
        registerCustomHelpers(env)

        findAnnotatedClasses(env)
        registerHelpersFromMockAnnotatedClasses()

        generateMockClasses()

        return false
    }

    private fun generateMockClasses() {
        try {
            mockClasses.values.forEach { MockGenerator.generateClass(mElementUtils, mTypes, mFiler, mMessager, it) }
        } catch (anyEx: Exception) {
            e(null, anyEx.toString())
        }

        mockClasses.clear()
    }

    private fun findAnnotatedClasses(env: RoundEnvironment) {
        env.getElementsAnnotatedWith(Mock::class.java)
                .filter { it.isValidAnnotatedElement() }
                .forEach { registerMockAnnotatedClass(it) }
    }

    private fun registerHelpersFromMockAnnotatedClasses() {
        mockClasses.keys.forEach {
            val fullClassName = mockClasses[it]!!.annotatedClassElement.toString()
            val postGeneratedFunction = Helper.fullGeneratedHelperPath(it)
            registeredFunctions[fullClassName] = postGeneratedFunction
        }
    }

    private fun registerMockAnnotatedClass(element: Element) {
        val annotatedClass = MockAnnotatedClass(element)

        if (mockClasses.containsKey(element.elementName())) {
        }

        mockClasses[element.elementName()] = annotatedClass
    }

    private fun registerDefaultHelpers() {
        val typeElement: TypeElement = mElementUtils.getTypeElement(MockUtils::javaClass.name)
        registerFunctions(typeElement.enclosedElements)
    }

    private fun registerCustomHelpers(env: RoundEnvironment) {
        registerFunctions(env.getElementsAnnotatedWith(MockTypeHelper::class.java))
    }

    fun registerFunctions(elements: Collection<out Element>) {
        elements
                .filter { it.isMockHelperMethodValid() }
                .forEach { registerFunction(it, it.helperFunctionName()) }
    }

    fun registerFunction(element: Element, helperFunctionName: String) {
        try {
            val helper: MockTypeHelper = element.getAnnotation(MockTypeHelper::class.java)
            helper.types.forEach { registeredFunctions.put(it.toString(), helperFunctionName) }
        } catch (typesEx: MirroredTypesException) {
            typesEx.typeMirrors.forEach { registeredFunctions.put(it.toString(), helperFunctionName) }
        }
    }

    private fun e(e: Element?, msg: String, vararg args: Any) {
        mMessager.printMessage(
                Diagnostic.Kind.ERROR,
                String.format(msg, *args),
                e)
    }

    override fun getSupportedAnnotationTypes(): MutableSet<String> {
        return mutableSetOf<String>().apply {
            add(Mock::javaClass.name)
            add(MockTypeHelper::javaClass.name)
        }
    }

    override fun getSupportedSourceVersion(): SourceVersion {
        return SourceVersion.latestSupported()
    }

    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
    // HELP METHODS
    ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

    fun Element?.elementName(): String {
        return checkNotNull(this).simpleName.toString()
    }

    fun Element?.isMockHelperMethodValid(): Boolean {
        checkNotNull(this)

        val isMethod = this!!.kind == ElementKind.METHOD
        val isPublic = this.modifiers.contains(Modifier.PUBLIC)
        val isStatic = this.modifiers.contains(Modifier.STATIC)

        return isMethod && isPublic && isStatic
    }

    fun Element?.helperFunctionName(): String {
        checkNotNull(this)
        checkNotNull(this!!.enclosingElement)

        val className = this.enclosingElement!!.asType().toString()
        val functionName = this.toString()

        return "$className.$functionName"
    }

    fun Element.isValidAnnotatedElement(): Boolean {
        var isValid = true

        if (this.kind != ElementKind.CLASS) {
            e(this, "Only classes can be annotated with @${Mock::javaClass.name}")
            isValid = false
        }
        if (this.modifiers.contains(Modifier.PRIVATE)) {
            e(this, "Private class can`t be annotated with @${Mock::javaClass.name}")
            isValid = false
        }

        return isValid
    }
}